### PR TITLE
Topic/blaze client cleanup

### PR DESCRIPTION
**Pull request by bryce-anderson**
This cleans up a lot of the OO patterns that were used to build the blaze-client package.

### CODE REVIEW

File: blaze-client/src/main/scala/org/http4s/client/blaze/ConnectionBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+import scalaz.concurrent.Task
+
+trait ConnectionBuilder {
+
+  /** Free resources associated with this client factory */
+  def shutdown(): Task[Unit]
+
+  /** Attempt to make a new client connection */
+  def makeClient(req: Request): Future[BlazeClientStage]
```
Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/client/blaze/ConnectionBuilder.scala]: This API is exposing both a Task and a Future.  Task is more in line with the rest of the API.

File: blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala [hunk 2]
```diff
@@ -4,32 +4,17 @@ import org.http4s.blaze.pipeline.Command
...
-  protected def recycleClient(request: Request, stage: BlazeClientStage): Unit = stage.shutdown()
-
-  /** Get a connection to the provided address
-    * @param request [[Request]] to connect too
-    * @param fresh if the client should force a new connection
-    * @return a Future with the connected [[BlazeClientStage]] of a blaze pipeline
-    */
-  protected def getClient(request: Request, fresh: Boolean): Future[BlazeClientStage]
-
+final class BlazeClient(manager: ConnectionManager, ec: ExecutionContext) extends Client {
```
Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala]: I would see this taking an implicit ExecutorService.  If you need an ExecutionContext in the bowels, I have a conversion on a branch.

Comment (bryce-anderson) [inline blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala]: Its easy to get an ```ExecutionContext``` from an ```ExecutorService```, its the other way around thats a bit tricky.

Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala]: Oh, that's right.  Well, I have the harder direction laying around somewhere.

**Pull request merged**